api/client/identityfile.go

```diff
@@ -0,0 +1,213 @@
+/*
+Copyright 2021 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"bufio"
+	"bytes"
+	"crypto/tls"
+	"crypto/x509"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+
+	"github.com/gravitational/trace"
+)
+
+const (
+	// IdentityFilePermissions defines file permissions for Identity files.
+	IdentityFilePermissions = 0600
+)
+
+// IdentityFile represents the basic components of an identity file.
+type IdentityFile struct {
+	// PrivateKey is a PEM encoded key.
+	PrivateKey []byte
+	Certs      Certs
+	CACerts    CACerts
+}
+
+// Certs contains PEM encoded certificates.
+type Certs struct {
+	// SSH is a cert used for SSH.
+	SSH []byte
+	// TLS is a cert used for TLS.
+	TLS []byte
+}
+
+// CACerts contains PEM encoded CA certificates.
+type CACerts struct {
+	// SSH are CA certs used for SSH.
+	SSH [][]byte
+	// TLS are CA certs used for TLS.
+	TLS [][]byte
+}
+
+// TLS returns the identity file's associated TLS config.
+func (i *IdentityFile) TLS() (*tls.Config, error) {
+	cert, err := tls.X509KeyPair(i.Certs.TLS, i.PrivateKey)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	pool := x509.NewCertPool()
+	for _, caCerts := range i.CACerts.TLS {
+		if !pool.AppendCertsFromPEM(caCerts) {
+			return nil, trace.BadParameter("invalid CA cert PEM")
+		}
+	}
+
+	return &tls.Config{
+		Certificates: []tls.Certificate{cert},
+		RootCAs:      pool,
+	}, nil
+}
+
+// WriteIdentityFile writes the given identityFile to the specified path.
+func WriteIdentityFile(idFile *IdentityFile, path string) error {
+	buf := new(bytes.Buffer)
+	if err := encodeIdentityFile(buf, idFile); err != nil {
+		return trace.Wrap(err)
+	}
+	err := ioutil.WriteFile(path, buf.Bytes(), IdentityFilePermissions)
+	return trace.Wrap(err)
```

Comment:
nit: `trace.ConvertSystemError` which perform additional mapping of stdlib errors to trace type space.